### ISSUE TITLE
Allow `A` and `B` to be the same type

### ISSUE THREAD

**Issue by eschnett** [andyferris/Switches.jl#2]
I think your constructor doesn't work when `A` and `B` are the same type, since you are currently dispatching based on the argument type of the constructor.

This might seem irrelevant at first (why would you need a union if both types are the same?), but the point of the tag is that a tagged union still "knows" which case it is handling since the tag is still present. Technically speaking, a tagged union is a disjoint union of its elements, not the (standard) set union.

**Comment by andyferris** [comment]
True, Erik. I still consider this kind of behaviour "not finished" in terms of what makes the most sense.

Primarily, I would like to implement a `Tuple` based version where you can have any number of different types (including just 1) and maybe a `UInt8` to indicate which, so the semantics are a bit nicer. 

On the other hand, I thought some nice syntax would be:
```julia
s = Switch{Int, Float64}(1)
if hastype(s, Int)
    val = s[Int]
    # do something
else
    val = s[Float64]
    # do something else
end
```
so `getindex` is based on the type that is stored (and `@inbounds` can indicate if you've already safely checked the type). This would make it a kind of associative container from types to values. This wouldn't work if the type wasn't unique... so I'm not sure how to trade off things like this!

**Comment by eschnett** [comment]
I guess there are different use cases to consider. From an abstract point of view, an associative container from types to values -- i.e. a more efficient `Union` -- makes sense.

What I've been looking for is different. I usually start from a case distinction where I have something like an enum. For each enum value there is a (potentially) different type. Since the enums are fundamental, the types might all be the same. When creating such an object, I know what case this is (i.e. I have the respective enum value), and thus I'd use a constructor that takes two arguments, an enum (probably even encoded in `Val{}` for compile-time dispatch) and the value I want to store.

**Comment by andyferris** [comment]
OK, I see. 

I wonder if there are other nice ways of achieving this... if I understand you correctly, here's just one pattern:
```julia
abstract EnumClass

immutable A <: EnumClass; end
immutable B <: EnumClass; end

enumvalue(::A) = 1
enumvalue(::B) = 2.0
```
You can write functions that accept an `EnumClass` and they will automatically specialize the correct call to `enumvalue()`, and you can use `Switch{A,B}` as a container if you need to for whatever reason.
